File: SymmetricTreeBFS.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
//tc- o(n)
//sc-o(n)
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if(root == null) return false;
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root.left);
        q.add(root.right);
        
        while(!q.isEmpty())
        {
            TreeNode left = q.poll();
            TreeNode right = q.poll();
            if(left==null && right == null) continue;
            if(left == null || right == null || left.val != right.val) return false;
            
            q.add(left.left);
            q.add(right.right);
            q.add(left.right);
            q.add(right.left);
            
        }
        return true;
        
}
}

File: SymmetricTreeBFS.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    
    '''Time Complexity: O(n)
    Space Complexity: O(h)'''
  
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        #Using bfs
        if not root:
            return True
        
        #Create queue... no need to check to check for the root
        queue=deque()
        #add left and right child of the root
        queue.append(root.left)
        queue.append(root.right)
        #while queue is not empty
        while len(queue)!=0:
            #pop fifo
            left=queue.popleft()
            right=queue.popleft()
            #if both none continue
            if left is None and right is None:
                continue
            #else return false
            if left is None or right is None or left.val != right.val:
                return False
            #add
            queue.append(left.left)
            queue.append(right.right)
            queue.append(left.right)
            queue.append(right.left)
        return True
            
        
        
